src/pybind/feat/wave_reader_pybind.cc

```diff
@@ -0,0 +1,83 @@
+// pybind/feat/wave_reader_pybind.cc
+
+// Copyright 2019   Microsoft Corporation (author: Xingyu Na)
+
+// See ../../COPYING for clarification regarding multiple authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
+// WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+// See the Apache 2 License for the specific language governing permissions and
+// limitations under the License.
+
+#include "feat/wave_reader_pybind.h"
+#include "feat/wave-reader.h"
+#include "util/table-types.h"
+
+using namespace kaldi;
+
+void pybind_wave_reader(py::module& m) {
+  m.attr("kWaveSampleMax") = py::cast(kWaveSampleMax);
+
+  py::class_<WaveInfo>(m, "WaveInfo")
+      .def("IsStreamed", &WaveInfo::IsStreamed)
+      .def("SampFreq", &WaveInfo::SampFreq)
+      .def("SampleCount", &WaveInfo::SampleCount)
+      .def("Duration", &WaveInfo::Duration)
+      .def("NumChannels", &WaveInfo::NumChannels)
+      .def("BlockAlign", &WaveInfo::BlockAlign)
+      .def("DataBytes", &WaveInfo::DataBytes);
+
+  py::class_<WaveData>(m, "WaveData")
+      .def("Duration", &WaveData::Duration)
+      .def("Data", &WaveData::Data);
```

Comment:
```
.def("Data", &WaveData::Data, py::return_value_policy::reference)
```
python does not own the returned object here, so we have to use reference semantics.

Reply:
I want to get things clear here. The returned type is MatrixBase<Float>, which is bound in matrix module, which use non-copy policy. So if the ownership is taken by Python, the wave data will be garbage-collected if it is owned by a local variable when running out of its life time. Right?

Reply:
MatrixBase being destroyed would not lead to the underlying data being destroyed, probably just the MatrixBase object itself.  But reference semantics may still make more sense.

Reply:
I have just tested it locally


```python
.def(py::init<BaseFloat, const MatrixBase<BaseFloat>&>(), py::arg("sample_fre"), py::arg("data"))
.def("Data", &WaveData::Data, py::return_value_policy::reference)
.def("Data2", &WaveData::Data)
```

And in Python

```python
import kaldi
m = kaldi.FloatMatrix(2, 3)
wave_data = kaldi.WaveData(sample_fre=16000, data=m)

d = wave_data.Data()
d2 = wave_data.Data2()

assert id(d) == id(d2) == id(wave_data.Data()) == id(wave_data.Data2())
assert isinstance(d, kaldi.FloatMatrix)
assert isinstance(d2, kaldi.FloatMatrix)
```

It indeed works without `py::return_value_policy::reference` in this case.

But I will still use `py::return_value_policy::reference` since it returns a reference.

Reply:
https://github.com/danpovey/kaldi/blob/532f3845a804e85842670a6a248d1db3aac0c32f/src/matrix/kaldi-matrix.h#L743-L750

The constructor and destructor of `MatrixBase` are protected. You can never create an object of `MatrixBase` directly. So the returned type is a subclass of `MatrixBase`